<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Ralph">
  <title>Spatial population genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="reveal.js/css/reveal.css">
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <link rel="stylesheet" href="reveal.js/css/theme/simple.css" id="theme">
  <!-- Printing and PDF exports -->
  <script>
    var link = document.createElement( 'link' );
    link.rel = 'stylesheet';
    link.type = 'text/css';
    link.href = window.location.search.match( /print-pdf/gi ) ? 'reveal.js/css/print/pdf.css' : 'reveal.js/css/print/paper.css';
    document.getElementsByTagName( 'head' )[0].appendChild( link );
  </script>
  <!--[if lt IE 9]>
  <script src="reveal.js/lib/js/html5shiv.js"></script>
  <![endif]-->
  <style type="text/css">
  
  .reveal { font-size: 30px; }
  
  .reveal h1 { font-size: 1.5em; } 
  
  .reveal h2 { font-size: 1.2em; } 
  
  .reveal .slides { text-align: left; }
  
  .reveal .slides figure { text-align: center; }
  
  .centered p { text-align: center; }
  
  .reveal figcaption { display: none; }
  
  .reveal img { border: none; }
  
  .reveal section img { border: none; }
  
  .MathJax_Display { margin: 0; }
  
  </style>
  <div style="display: none">
  \[
  \]
  </div>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Spatial population genetics</h1>
  <p class="subtitle">ecology, evolution, and simulation</p>
  <p class="author">Peter Ralph</p>
  <p class="date">Moscow, ID // March 2019</p>
</section>

<section><section id="overview" class="title-slide slide level1"><h1>Overview</h1></section><section id="section" class="slide level2">
<h2></h2>
<ol type="1">
<li><p>Two problems, about geography and evolution</p></li>
<li><p>Tree sequences</p>
<ul>
<li>for more and better simulations</li>
</ul></li>
<li><p>Some steps forward</p></li>
</ol>
<!-- Tortoise intro -->
</section></section>
<section><section id="problem-1" class="title-slide slide level1"><h1>Problem #1</h1></section><section id="the-mojave-desert-tortoise" class="slide level2">
<h2>the Mojave Desert Tortoise</h2>
<p><em>Gopherus agassizii:</em></p>
<div class="centered">
<figure>
<img data-src="figs/tortoise/tortoise-in-burrow.jpg" alt="a tortoise" style="width:70.0%" /><figcaption>a tortoise</figcaption>
</figure>
</div>
</section><section id="the-mojave-desert-tortoise-1" class="slide level2">
<h2>the Mojave Desert Tortoise</h2>
<div class="centered">
<figure>
<img data-src="figs/tortoise/range-abundance-map.jpeg" alt="range map" style="width:70.0%" /><figcaption>range map</figcaption>
</figure>
</div>
</section><section id="the-mojave-desert-tortoise-2" class="slide level2">
<h2>the Mojave Desert Tortoise</h2>
<div class="centered">
<figure>
<img data-src="figs/tortoise/ivanpah-opens.png" alt="ivanpah" style="width:70.0%" /><figcaption>ivanpah</figcaption>
</figure>
</div>
</section><section id="the-mojave-desert-tortoise-3" class="slide level2">
<h2>the Mojave Desert Tortoise</h2>
<div class="centered">
<figure>
<img data-src="figs/tortoise/latimes-torts-delay-solar.png" alt="solar" style="width:70.0%" /><figcaption>solar</figcaption>
</figure>
</div>
</section><section id="the-questions" class="slide level2">
<h2>The question(s)</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li><p>How will changes to the landscape affect population viability and gene flow?</p></li>
<li><p>How do tortoises move around on the landscape?</p></li>
</ol>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/tortoise/drecp-pref-alt-snapshot.png" /></p>
</div>
</div>
<!-- Mimulus intro -->
</section></section>
<section><section id="problem-2" class="title-slide slide level1"><h1>Problem #2</h1></section><section id="genomic-landscapes" class="slide level2">
<h2>Genomic landscapes</h2>
<figure>
<img data-src="figs/from_the_literature/langley-et-al-2012-chr3-pi-and-rho.png" alt="Langley et al 2012" /><figcaption>Langley et al 2012</figcaption>
</figure>
</section><section id="section-1" class="slide level2">
<h2></h2>
<p>What causes variation in genetic diversity?</p>
</section><section id="section-2" class="slide level2">
<h2></h2>
<p><span class="math display">\[
\begin{aligned}
\pi &amp;= \text{ (within-pop diversity) } \\
d_{xy} &amp;= \text{ (between-pop divergence) } \\
F_{ST} &amp;= \text{ (relative differentiation) }
\end{aligned}
\]</span></p>
<div class="centered">
<p><img data-src="figs/fst_tree.png" style="width:60.0%" /></p>
</div>
</section><section id="diversity-correlates-with-recombination-rate" class="slide level2">
<h2>Diversity correlates with recombination rate</h2>
<figure>
<img data-src="figs/from_the_literature/corbett-detig-divergence-recomb-drosophila.png" alt="Corbett-Detig et al" /><figcaption>Corbett-Detig et al</figcaption>
</figure>
<p>Corbett-Detig et al 2015; Cutter &amp; Payseur 2013</p>
</section><section class="slide level2">

<figure>
<img data-src="figs/from_the_literature/corbett-detig-divergence-recomb-all-species.png" alt="Corbett-Detig et al" style="width:70.0%" /><figcaption>Corbett-Detig et al</figcaption>
</figure>
<p>Corbett-Detig et al 2015</p>
<!--

-------------------

![Langley et al 2012](figs/from_the_literature/langley-et-al-2012-chr3-pi-and-rho.png)

-------------------

![McVicker et al 2009](figs/from_the_literature/mcvicker-2009-hu_chr1-landscape.png)
![](figs/from_the_literature/mcvicker-2009-hu_chr1-landscape-caption.png)

McVicker et al 2009
-->
</section><section class="slide level2">

<dl>
<dt>linked selection</dt>
<dd><p>The indirect effects of selection on genomic locations that are <em>linked</em> to the sites under selection by a lack of recombination.</p>
</dd>
</dl>
<div class="centered">
<p><img data-src="figs/sweep_haps.png" class="fragment" style="width:40.0%" /> <img data-src="figs/background_haps.png" class="fragment" style="width:40.0%" /></p>
</div>
</section><section id="the-mimulus-aurantiacus-species-complex" class="slide level2">
<h2>The <em>Mimulus aurantiacus</em> species complex</h2>
<div class="centered">
<p><img data-src="figs/aurantiacus/phylogeny.png" /></p>
</div>
<!--

------------------

::: {.centered}
![](figs/aurantiacus/stankowski-color-cline.png){width=70%}
:::

-->
</section><section class="slide level2">

<figure>
<img data-src="figs/aurantiacus/preprint.png" alt="https://www.biorxiv.org/content/early/2018/06/21/342352" /><figcaption>https://www.biorxiv.org/content/early/2018/06/21/342352</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/aurantiacus/preprint_peeps.png" alt="https://www.biorxiv.org/content/early/2018/06/21/342352" /><figcaption>https://www.biorxiv.org/content/early/2018/06/21/342352</figcaption>
</figure>
</section><section class="slide level2">

<p>The data:</p>
<ul>
<li>chromosome-level genome assembly</li>
<li><span class="math inline">\(20\times\)</span> coverage of 8 taxa and outgroup (<em>M.clevelandii</em>)</li>
<li>diversity (<span class="math inline">\(\pi\)</span>), divergence (<span class="math inline">\(d_{xy}\)</span>), and differentiation (<span class="math inline">\(F_{ST}\)</span>) in windows</li>
<li>36 pairwise comparisons among 9 taxa</li>
<li>estimates of recombination rate and gene density from map and annotation</li>
</ul>
<!--
## A spectrum of differentiation

![](figs/aurantiacus/distributions.png)
-->
</section><section class="slide level2">

<p><img data-src="figs/aurantiacus/lg1.png" /></p>
</section><section class="slide level2">

<div class="centered">
<p><img data-src="figs/aurantiacus/dxy-pi-cor-by-time.png" style="width:60.0%" /></p>
</div>
</section><section id="section-3" class="slide level2">
<h2></h2>
<ul>
<li><p>Emergence of landscape of diversity across <span class="math inline">\(\approx\)</span> 1.5 million years!</p></li>
<li><p>Shared targets of linked selection across taxa?</p></li>
</ul>
<div class="fragment">
<ul>
<li>But, what <em>kind</em> of linked selection?<br />
How strongly and frequently does it act?<br />
On how many targets?</li>
</ul>
</div>
<div class="fragment">
<p>Consequences for: genetic load/disease, speed of evolution, trait architecture.</p>
<!-- section SIMULATION -->
</div>
</section></section>
<section><section id="simulation-not-as-easy-as-you-might-think" class="title-slide slide level1"><h1>Simulation: not as easy as you might think</h1></section><section id="simulations-to" class="slide level2">
<h2>Simulations, to</h2>
<ul>
<li class="fragment"><p>make predictions</p></li>
<li class="fragment"><p>guide theory</p></li>
<li class="fragment"><p>train inference methods</p></li>
</ul>
</section><section class="slide level2">

<p>We need:</p>
<ul>
<li>large populations with long genomes</li>
<li>many loci under selection</li>
<li>geographic population structure</li>
</ul>
<p>for long enough to reach equilibrium (<span class="math inline">\(\sim 10N\)</span> generations).</p>
<div class="fragment">
<p>For instance: <span class="math inline">\(10^4\)</span> individuals with <span class="math inline">\(2 \times 10^8\)</span> base pairs each for <span class="math inline">\(10^5\)</span> generations; nonneutral mutations possible at <span class="math inline">\(10^5\)</span> loci.</p>
</div>
</section><section id="section-4" class="slide level2">
<h2></h2>
<p><em>First:</em></p>
<ul>
<li><strong>large populations with long genomes</strong></li>
</ul>
<!-- 1. explain tree sequences and why they are so efficient (5min) -->
</section></section>
<section><section id="the-tree-sequence" class="title-slide slide level1"><h1>The tree sequence</h1></section><section class="slide level2">

<p><img data-src="figs/treseq_preprint.png" /></p>
</section><section class="slide level2">

<p><img data-src="figs/treseq_preprint_peeps.png" /></p>
</section><section id="history-is-a-sequence-of-trees" class="slide level2">
<h2>History is a sequence of trees</h2>
<p>For a set of sampled chromosomes, at each position along the genome there is a genealogical tree that says how they are related.</p>
<div class="fragment">
<figure>
<img data-src="figs/sim_ts.anim.gif" alt="Trees along a chromosome" /><figcaption>Trees along a chromosome</figcaption>
</figure>
</div>
</section><section class="slide level2">

<p>A <strong>tree sequence</strong> describes this, er, sequence of trees.</p>
<!--
. . .

*Observations:*

1. The *pedigree* (parental relationships) plus crossover locations
    would give us the tree sequence for *everyone, ever*.

2. Much less can fully describe the history relevant to a *sample* of genomes.

3. This information is equivalent to the Ancestral Recombination Graph (ARG).
-->
</section><section class="slide level2">

<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, and McVean</a> introduced the <strong>tree sequence</strong> data structure for a fast coalescent simulator, <a href="https://github.com/jeromekelleher/msprime">msprime</a>.</p>
<ul>
<li><p>stores sequence <em>and</em> genealogical data</p></li>
<li><p>tree differences allow <strong>very</strong> efficient</p>
<ul>
<li>subsetting</li>
<li>calculation of allele frequencies in arbitrary cohorts</li>
<li>linkage disequilibrium</li>
<li>log-time haplotype matching</li>
</ul></li>
<li><p>tree-based sequence storage closely related to haplotype-matching compression</p></li>
</ul>
</section><section id="simulated-file-sizes" class="slide level2">
<h2>Simulated file sizes</h2>
<div class="centered">
<figure>
<img data-src="figs/tsinfer_sizes.png" alt="file sizes" style="width:90.0%" /><figcaption>file sizes</figcaption>
</figure>
</div>
<p>from <em>Kelleher et al 2018, </em>Inferring the ancestry of everyone"*</p>
<!-- Estimated sizes of files required to store the genetic variation data for a
simulated human-like chromosome (100 megabases) for up to 10 billion haploid
(5 billion diploid) samples. Simulations were run for 10 1 up to 10 7 haplotypes
using msprime [Kelleher et al., 2016], and the sizes of the resulting files plotted
(points). -->
</section><section id="example-three-samples-two-trees-two-variant-sites" class="slide level2">
<h2>Example: three samples; two trees; two variant sites</h2>
<figure>
<img data-src="figs/example_tree_sequence.png" alt="Example tree sequence" /><figcaption>Example tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<p>Storing a tree sequence in the four tables - <em>nodes</em>, <em>edges</em>, <em>sites</em>, and <em>mutations</em> - is <em>succinct</em> (no redundancy).</p>
<div class="fragment">
<p>These are stored efficiently on disk with a bit more information (e.g., metadata).</p>
</div>
</section><section id="nodes-and-edges" class="slide level2">
<h2>Nodes and edges</h2>
<dl>
<dt>Edges</dt>
<dd><p>Who inherits from who; only <em>necessary</em> for coalescent events.</p>
<p>Records: interval (left, right); parent node; child node.</p>
</dd>
<dt>Nodes</dt>
<dd><p>The ancestors those happen in.</p>
<p>Records: time ago (of birth); ID (implicit).</p>
</dd>
</dl>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.0.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.1.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.2.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.3.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.4.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.5.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/nodes_edges_walkthrough/nodes_edges_walkthrough.6.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section id="sites-and-mutations" class="slide level2">
<h2>Sites and mutations</h2>
<dl>
<dt>Mutations</dt>
<dd><p>When state changes along the tree.</p>
<p>Records: site it occured at; node it occurred in; derived state.</p>
</dd>
<dt>Sites</dt>
<dd><p>Where mutations fall on the genome.</p>
<p>Records: genomic position; ancestral (root) state; ID (implicit).</p>
</dd>
</dl>
</section><section class="slide level2">

<figure>
<img data-src="figs/sites_muts_walkthrough/sites_muts_walkthrough.0.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/sites_muts_walkthrough/sites_muts_walkthrough.1.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/sites_muts_walkthrough/sites_muts_walkthrough.2.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/sites_muts_walkthrough/sites_muts_walkthrough.3.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/sites_muts_walkthrough/sites_muts_walkthrough.4.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
<!-- 4. explain application to fwds sim recording (5min) -->
</section></section>
<section><section id="forwards-time-tree-sequence-recording" class="title-slide slide level1"><h1>Forwards-time tree sequence recording</h1></section><section id="the-main-idea" class="slide level2">
<h2>The main idea</h2>
<p>If we <em>record the tree sequence</em> that relates everyone to everyone else,</p>
<p>after the simulation is over we can put neutral mutations down on the trees.</p>
<div class="fragment">
<p>Since neutral mutations don’t affect demography,</p>
<p>this is <em>equivalent</em> to having kept track of them throughout.</p>
</div>
</section><section class="slide level2">

<p>This means recording the entire genetic history of <strong>everyone</strong> in the population, <strong>ever</strong>.</p>
<div class="fragment">
<p>It is <em>not</em> clear this is a good idea.</p>
</div>
</section><section id="tree-recording-strategy" class="slide level2">
<h2>Tree recording strategy</h2>
<p>Every time an individual is born, we must:</p>
<div>
<ol type="1">
<li class="fragment">add each contributing gamete to the Node Table,</li>
<li class="fragment">add entries to the Edge Table recording which parental copy each inherited each bit of genome from, and</li>
<li class="fragment">add any new selected mutations to the Mutation Table and (if necessary) their locations to the Site Table.</li>
</ol>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:15%;">
<figure>
<img data-src="figs/finger_right.png" alt="Rightarrow" style="width:100.0%" /><figcaption>Rightarrow</figcaption>
</figure>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:75%;">
<p>This produces <strong>waaaaay</strong> too much data.</p>
</div>
</div>
</div>
</section><section class="slide level2">

<p>We won’t end up needing the <em>entire</em> history of <em>everyone ever</em>,</p>
<div class="fragment">
<p>but we won’t know <em>what</em> we’ll need until later.</p>
</div>
<div class="fragment">
<p>How do we get rid of the extra stuff?</p>
</div>
</section><section id="simplification" class="slide level2">
<h2>Simplification</h2>
<p><em>Question:</em> given a tree sequence containing the history of many individuals, how do we <em>simplify</em> it to only the history of a subset?</p>
</section><section class="slide level2">

<p>Concretely, given an input tree sequence and a subset of its nodes we call the <em>samples</em>,</p>
<p>we want a new <em>minimal</em> tree sequence that contains the history of those samples.</p>
<!--
1. All marginal trees match the corresponding subtree 
    in the input tree sequence.

2. Every non-sample node in marginal trees has at least two children.

3. All nodes and edges are ancestral to at least one sample.

4. No adjacent redundant edges 
    (e.g., $(\ell, x, p, c) + (x, r, p, c) \rightarrow (\ell, r, p, c)$).
-->
</section><section class="slide level2">

<p>To simplify a tree sequence to the history of the <em>samples</em>:</p>
<ol type="1">
<li class="fragment"><p>Paint each <em>sampled</em> chromosome a distinct color.</p></li>
<li class="fragment"><p>Moving back up the tree sequence, copy colors of each chromosome to the parental chromosomes they inherited from.</p></li>
<li class="fragment"><p>If two colors go in the same spot (<em>coalescence</em>), replace with a new color (unique to that ancestor). Output a node for the ancestor and an edge for the coalescence.</p></li>
<li class="fragment"><p>Once all colors have coalesced in a given segment, stop propagating it.</p></li>
</ol>
</section><section id="an-example-simplify-these-to-j-and-k" class="slide level2">
<h2>An example: simplify these to J and K</h2>
<figure>
<img data-src="figs/simplify_walkthrough.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.0.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.1.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.2.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.3.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.4.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.5.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.6.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.7.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.8.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.9.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.10.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.11.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.12.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.13.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.14.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/simplify_walkthrough/simplify_walkthrough.15.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section id="wright-fisher-n10-before-simplification" class="slide level2">
<h2>Wright-Fisher, N=10: before simplification</h2>
<figure>
<img data-src="figs/sim_wf.anim.gif" alt="Wright-Fisher tree sequence" /><figcaption>Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="wright-fisher-n10-before-simplification-1" class="slide level2">
<h2>Wright-Fisher, N=10: before simplification</h2>
<figure>
<img data-src="figs/sim_wf_unlabeled.anim.gif" alt="Wright-Fisher tree sequence" /><figcaption>Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="and-after-simplification" class="slide level2">
<h2>… and after simplification</h2>
<figure>
<img data-src="figs/sim_wf_simplified.anim.gif" alt="Simplified Wright-Fisher tree sequence" /><figcaption>Simplified Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="revised-tree-recording-strategy" class="slide level2">
<h2><em>Revised</em> tree recording strategy</h2>
<p>Every time an individual is born, we must:</p>
<ol type="1">
<li>add each gamete to the Node Table,</li>
<li>add entries to the Edge Table recording which parent each gamete inherited each bit of genome from</li>
<li>add any new mutations to the Mutation Table and (if necessary) their locations to the Site Table.</li>
</ol>
<p>… and,</p>
<ol start="4" type="1">
<li>Every so often, <em>simplify</em> the tables so far, retaining the history of the current generation.</li>
</ol>
</section></section>
<section><section id="implementation-and-results" class="title-slide slide level1"><h1>Implementation and results</h1></section><section id="benchmark-implementation" class="slide level2">
<h2>Benchmark implementation</h2>
<ul>
<li><p>Recording, simplifying, and output of tables: <code>C</code> code in <code>msprime</code>.</p></li>
<li><p>Simulation: <a href="https://github.com/molpopgen/fwdpp"><code>fwdpp</code></a>, by Kevin Thornton (in <code>C++</code>) (<a href="https://github.com/molpopgen/fwdpy11_arg_example">code</a>)</p></li>
<li><p>Glue: <a href="https://github.com/pybind/pybind11/"><code>pybind11</code></a> and <a href="http://www.numpy.org/"><code>numpy</code></a></p></li>
<li><p>Machine: Ubuntu / 2x 2.6 GHz Intel E5-2650 CPU</p></li>
</ul>
</section><section id="simulation-parameters" class="slide level2">
<h2>Simulation parameters</h2>
<ul>
<li><p>Wright-Fisher population of size <span class="math inline">\(N\)</span></p></li>
<li><p>simulated for <span class="math inline">\(10N\)</span> generations</p></li>
<li><p>neutral mutation rate <span class="math inline">\(\mu\)</span> equal to recombination rate <span class="math inline">\(r\)</span> per gamete</p></li>
<li><p>many, weakly deleterious mutations: rate <span class="math inline">\(\mu/100\)</span> with <span class="math inline">\(s\)</span> exponentially distributed with mean <span class="math inline">\(2.5/N\)</span>.</p></li>
</ul>
<div class="fragment">
<p><em>Note:</em> if we recorded tree sequences (“pedigree recording”) then the neutral mutation rate was <em>zero</em> but neutral mutations were added <em>afterwards</em>.</p>
</div>
</section><section class="slide level2">

<!--
Translation:
$$
\begin{aligned}
N=10^3 \text{ and } \rho=10^4 \\
\Rightarrow L \approx 2.5 \times 10^8 \text{bp}
\end{aligned}
$$

$$
\begin{aligned}
N=10^4 \text{ and } \rho=10^4 \\
\Rightarrow L \approx 2.5 \times 10^7 \text{bp}
\end{aligned}
$$

$$
\begin{aligned}
N=5 \times 10^4 \text{ and } \\
\qquad \rho=10^5 \\
\Rightarrow L \approx 5 \times 10^7 \text{bp}
\end{aligned}
$$
-->
<figure>
<img data-src="figs/rawspeed.png" alt="Total run time per single simulation as a function of region length." style="width:80.0%" /><figcaption>Total run time per single simulation as a function of region length.</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/speedup.png" alt="Relative speedup of simulations" style="width:80.0%" /><figcaption>Relative speedup of simulations</figcaption>
</figure>
</section></section>
<section><section id="geographic-simulations" class="title-slide slide level1"><h1>Geographic simulations</h1></section><section class="slide level2">

<p>To check predictions, we <strong>still</strong> need simulations with:</p>
<ul>
<li>many loci under selection</li>
<li><strong>geographic population structure</strong></li>
<li>large populations with long genomes</li>
</ul>
<p>for long enough to reach equilibrium (<span class="math inline">\(\sim 10N\)</span> generations).</p>
<div class="fragment">
<div class="columns">
<div class="column" style="width:15%;">
<figure>
<img data-src="figs/finger_right.png" alt="Rightarrow" style="width:100.0%" /><figcaption>Rightarrow</figcaption>
</figure>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:75%;">
<p>SLiM v3.1 <em>(Haller and Messer)</em></p>
</div>
</div>
</div>
</section><section id="section-5" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>S</strong>election on <strong>Li</strong>nked <strong>M</strong>utations, v3.1: a forwards-time, individual-based, explicitly genomic simulator, now with:</p>
<ul>
<li>continuous space</li>
<li>geographic maps</li>
<li>locally, density-dependent demography (<em>non-Wright-Fisher</em>)</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="figs/slim_logo.png" /> <!-- ![](figs/slim_space.png) --></p>
</div>
</div>
<div class="fragment">
<p><strong>And a GUI!!!</strong></p>
</div>
</section><section id="now-with-tree-sequence-recording" class="slide level2">
<h2>… now with tree sequence recording!</h2>
<p><img data-src="figs/slim_preprint.png" /></p>
</section><section id="a-100x-speedup" class="slide level2">
<h2>A 100x speedup:</h2>
<div class="centered">
<p><img data-src="figs/slim_timing.png" style="width:60.0%" /></p>
</div>
</section><section id="what-else-can-you-do-with-the-tree-sequence" class="slide level2">
<h2>What else can you do with the tree sequence?</h2>
<ul>
<li class="fragment">record ancient samples</li>
<li class="fragment">true ancestry reconstruction</li>
<li class="fragment"><em>recapitation</em>: fast, post-hoc initialization with coalescent simulation</li>
</ul>
</section><section class="slide level2">

<p><strong>For example:</strong></p>
<ul>
<li>genome as human chr7 (<span class="math inline">\(1.54 \times 10^8\)</span>bp)</li>
<li><span class="math inline">\(\approx\)</span> 10,000 diploids</li>
<li>500,000 overlapping generations</li>
<li>continuous, square habitat</li>
<li>selected mutations at rate <span class="math inline">\(10^{-10}\)</span></li>
<li>neutral mutations added afterwards</li>
</ul>
<p><em>Runtime:</em> 8 hours</p>
</section></section>
<section><section id="genomic-landscapes-1" class="title-slide slide level1"><h1>Genomic landscapes</h1></section><section class="slide level2">

<figure>
<img data-src="figs/aurantiacus/preprint.png" alt="https://www.biorxiv.org/content/early/2018/06/21/342352" /><figcaption>https://www.biorxiv.org/content/early/2018/06/21/342352</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/aurantiacus/preprint_peeps.png" alt="https://www.biorxiv.org/content/early/2018/06/21/342352" /><figcaption>https://www.biorxiv.org/content/early/2018/06/21/342352</figcaption>
</figure>
</section><section id="the-data" class="slide level2">
<h2>The data</h2>
<p><img data-src="figs/aurantiacus/phylogeny.png" /></p>
</section><section id="the-data-1" class="slide level2">
<h2>The data</h2>
<p><img data-src="figs/aurantiacus/divergence_by_node_LG1.png" /></p>
</section><section id="simulations" class="slide level2">
<h2>Simulations</h2>
<div class="columns">
<div class="column" style="width:70%;">
<ul>
<li><span class="math inline">\(N=10,000\)</span> diploids</li>
<li>burn-in for <span class="math inline">\(10N\)</span> generations</li>
<li><p>population split followed by period sampling, with:</p>
<ul>
<li>neutral</li>
<li>background selection</li>
<li>selection against introgressed alleles</li>
</ul></li>
</ul>
</div><div class="column" style="width:30%;">
<p><img data-src="figs/murillo.jpeg" /></p>
</div>
</div>
</section><section class="slide level2">

<p><img data-src="figs/aurantiacus/murillo_neutral.png" /></p>
</section><section class="slide level2">

<p><img data-src="figs/aurantiacus/murillo_selected.png" /></p>
<!-- Spaaaace results -->
</section></section>
<section><section id="spatial-landscapes" class="title-slide slide level1"><h1>Spatial landscapes</h1></section><section id="questions" class="slide level2">
<h2>Questions</h2>
<ol type="1">
<li><p>How does (continuous) space affect which population genetics signals?</p></li>
<li><p>Does it depend on <strong>sampling scheme</strong>?</p></li>
<li></li>
</ol>
</section></section>
    </div>
  </div>

  <script src="reveal.js/lib/js/head.min.js"></script>
  <script src="reveal.js/js/reveal.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        // Display the page number of the current slide
        slideNumber: true,
        // Push each slide change to the browser history
        history: true,
        // Transition style
        transition: 'none', // none/fade/slide/convex/concave/zoom
        math: {
          mathjax: '/usr/share/javascript/mathjax/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // Optional reveal.js plugins
        dependencies: [
          { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true },
          { src: 'reveal.js/plugin/math/math.js', async: true },
          { src: 'reveal.js/plugin/notes/notes.js', async: true }
        ]
      });
    </script>
    </body>
</html>
